File: akka-essentials/Chapter3/ActorMessagingExample/src/main/scala/org/akka/essentials/future/example/ProcessOrderActor.scala
package org.akka.essentials.future.example

import _root_.java.util.concurrent.TimeUnit
import akka.actor.Actor
import akka.actor.Props
import akka.pattern.ask
import akka.pattern.pipe
import akka.util.Timeout
import scala.concurrent.Future
import scala.concurrent.duration.Duration

class ProcessOrderActor extends Actor {

    //import the implicit

    import context._

    implicit val timeout = Timeout(5, TimeUnit.SECONDS)
    val orderActor = context.actorOf(Props[OrderActor])
    val addressActor = context.actorOf(Props[AddressActor])
    val orderAggregateActor = context.actorOf(Props[OrderAggregateActor])

    def receive = {
        case userId: Integer =>
            val aggResult: Future[OrderHistory] =
                for {
                    order <- ask(orderActor, userId).mapTo[Order] // call pattern directly
                    address <- addressActor ask userId mapTo manifest[Address] // call by implicit conversion
                } yield OrderHistory(order, address)

            aggResult pipeTo orderAggregateActor
    }
}

File: akka-essentials/Chapter3/ActorMessagingExample/src/main/java/org/akka/essentials/future/example/java/ProcessOrderActor.java
package org.akka.essentials.future.example.java;

import static akka.pattern.Patterns.ask;
import static akka.pattern.Patterns.pipe;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.concurrent.TimeUnit;

import org.akka.essentials.future.example.messages.Address;
import org.akka.essentials.future.example.messages.Order;
import org.akka.essentials.future.example.messages.OrderHistory;

import akka.actor.ActorRef;
import akka.actor.Props;
import akka.actor.UntypedActor;
import scala.concurrent.Future;
import akka.dispatch.Futures;
import akka.dispatch.Mapper;
import akka.util.Timeout;
import scala.concurrent.duration.Duration;

public class ProcessOrderActor extends UntypedActor {

    final Timeout t = new Timeout(Duration.create(5, TimeUnit.SECONDS));
    ActorRef orderActor = getContext().actorOf(Props.create(OrderActor.class));
    ActorRef addressActor = getContext().actorOf(Props.create(AddressActor.class));
    ActorRef orderAggregateActor = getContext().actorOf(Props.create(OrderAggregateActor.class));

    @Override
    public void onReceive(Object message) throws Exception {

        if (message instanceof Integer) {
            Integer userId = (Integer) message;
            final ArrayList<Future<Object>> futures = new ArrayList<Future<Object>>();
            //make concurrent calls to actors
            futures.add(ask(orderActor, userId, t));
            futures.add(ask(addressActor, userId, t));

            //set the sequence in which the reply are expected
            final Future<Iterable<Object>> aggregate = Futures.sequence(
                    futures, getContext().system().dispatcher());

            //once the replies comes back, we loop through the Iterable to
            //get the replies in same order
            final Future<OrderHistory> aggResult = aggregate
                    .map(new Mapper<Iterable<Object>, OrderHistory>() {
                        public OrderHistory apply(Iterable<Object> coll) {
                            final Iterator<Object> it = coll.iterator();
                            final Order order = (Order) it.next();
                            final Address address = (Address) it.next();
                            return new OrderHistory(order, address);
                        }
                    }, getContext().system().dispatcher());
            //aggregated result is piped to another actor
            pipe(aggResult, getContext().system().dispatcher()).to(orderAggregateActor);
        }
    }
}
